### ISSUE TITLE
Support shared library on web client applications

### ISSUE THREAD

**Issue by jurevans** [anoma/namada#373]
# Discussion and request

For client consumers of the `namada/shared` libraries, web clients will need to be able to compile via `wasm-pack` to a web target (and to a NodeJS target for running tests).

It would also be very useful to pursue the goal of handling the packaging of the `namada/shared` functionality for web-based clients. This could potentially be implemented by exposing any needed functionality by the client via `#[wasm-bindgen]` macros, executing `wasm-pack build` during the release process, versioning it and publishing the resulting NPM package to a Github repository that we could link to by version. This would enable any web client to import via a `package.json` and build and test against that version.

### Summary

- It would be ideal to be able to build `namada/shared` to a web-wasm target via `wasm-pack build --target [web|nodejs]` <https://rustwasm.github.io/docs/wasm-pack/>
- It would greatly simplify the process of importing functionality from `namada/shared` if top-level functionality is exposed via [wasm-bindgen](https://rustwasm.github.io/wasm-bindgen) (to TypeScript type definitions/JavaScript/wasm)
- During a release the new version could also generated as an NPM package via `wasm-pack build` and published to a repo we can link to (via e.g. `namada-interface` and future clients)

**Comment by tzemanovic** [comment]
At high level this sounds - we'll need to do more refactoring to expose more stuff, but I agree which should aim for something like this. Maybe we can use something like `#[cfg_attr(feature = "js-bindgen", wasm_bindgen)]` for conditional compilation on it.

For MASP, I think the WASM incompatibilities have been sorted out. @Scar26 actually started with some refactors in https://github.com/anoma/namada/pull/357 that should get us closer to having this re-usable (https://github.com/anoma/namada/pull/357/commits/6eeb22410c2a94ae0ff97afba3e31fc4553d0b89 and https://github.com/anoma/namada/pull/357/commits/d315ab073ca93474915a0b21cb897681fa54935a).

**Comment by jurevans** [comment]
Sounds good!

> For MASP, I think the WASM incompatibilities have been sorted out. @Scar26 actually started with some refactors in #357 that should get us closer to having this re-usable ([6eeb224](https://github.com/anoma/namada/commit/6eeb22410c2a94ae0ff97afba3e31fc4553d0b89) and [d315ab0](https://github.com/anoma/namada/commit/d315ab073ca93474915a0b21cb897681fa54935a)).

That's excellent! We will start an issue on the interface to begin removing some of our work-arounds to incorporate the MASP changes and see if we can clean things up. Thanks!

**Comment by cwgoes** [comment]
@murisi I think this falls under the scope of the shared crate work?

**Comment by murisi** [comment]
@cwgoes Yes I think this would fall in scope. But I agree with @jurevans that `wasm-pack` is perhaps a long-term goal. For now, @memasdeligeorgakis , @mateuszjasiuk , and I are focusing on the shortest/simplest way to get a working web client ready (in a manner that doesn't duplicate code from the CLI client)...

**Comment by cwgoes** [comment]
Closing as out-of-date.
